### PR TITLE
Prevent filename collisions 

### LINKED ISSUE #2171: Prevent filename collisions when importing images
## Description

If a user adds two objects with images that have the same filename, GDevelop will simply overwrite the file that was first added.

## Solution suggested

When importing assets, GDevelop should check if the filename already exists.  If it does, GDevelop should append a number that will ensure a unique filename.  This is similar to how Piskel appends numbers to edited images. 

## Demonstration of problem

In this video, I start from a new project and add two sprite objects with multiple frames each.   Source files for both objects have the same names, which GDevelop doesn't handle properly.

https://user-images.githubusercontent.com/8879811/103448259-34b25b80-4c54-11eb-9f96-3a0cd06e2d5b.mp4

### PR DESCRIPTION
**Pull request by rs4231199**
Fixes #2171

### CODE REVIEW

**Comment by 4ian** [comment]
Thanks for the PR! 
Can you explain a bit how this work, how you tested this, in which situation this helps? :)

File: newIDE/app/src/ResourcesList/ResourceUtils.js [hunk 1]
```diff
@@ -80,7 +80,26 @@ export const copyAllToProjectFolder = (
         return resourcePath;
       }
 
-      const resourceBasename = path.basename(resourcePath);
+      let resourceBasename = path.basename(resourcePath),
```
Comment (arthuro555) [inline newIDE/app/src/ResourcesList/ResourceUtils.js:83]: Try to keep what can be const const, like `splitAt` and `fileExtension`

File: newIDE/app/src/ResourcesList/ResourceUtils.js [hunk 2]
```diff
@@ -80,7 +80,26 @@ export const copyAllToProjectFolder = (
...
+        splitAt = resourceBasename.lastIndexOf('.'),
+        fileName = resourceBasename.substring(0, splitAt),
+        fileExtension = resourceBasename.substring(splitAt),
+        fileNumber = 1;
+
+      const AllResourcesNames = project
```
Comment (arthuro555) [inline newIDE/app/src/ResourcesList/ResourceUtils.js:89]: This is a normal variable, so this should be named in `camelCase`. `PascalCase` is used for classes only.

File: newIDE/app/src/ResourcesList/ResourceUtils.js [hunk 3]
```diff
@@ -80,7 +80,26 @@ export const copyAllToProjectFolder = (
...
+        if (fileName + fileExtension === AllResourcesNames.at(itr)) {
+          fileName = resourceBasename.substring(0, splitAt) + fileNumber;
+          fileNumber++;
+          itr = 0;
+        }
+      }
+
+      resourceBasename = fileName + fileExtension;
+
       const resourceNewPath = path.join(projectPath, resourceBasename);
```
Comment (arthuro555) [inline newIDE/app/src/ResourcesList/ResourceUtils.js:103]: No need to modify resourceBasename, you can do directly like this:
```suggestion
      const resourceNewPath = path.join(projectPath, fileName + fileExtension);
```
This makes the intention more clear and let you put `resourceBasename` as const to show it is a constant (the base name)

File: newIDE/app/src/ResourcesList/ResourceUtils.js [hunk 4]
```diff
@@ -80,7 +80,26 @@ export const copyAllToProjectFolder = (
...
+        fileExtension = resourceBasename.substring(splitAt),
+        fileNumber = 1;
+
+      const AllResourcesNames = project
+        .getResourcesManager()
+        .getAllResourceNames();
+
+      for (let itr = 0; itr < AllResourcesNames.size(); itr++) {
+        if (fileName + fileExtension === AllResourcesNames.at(itr)) {
+          fileName = resourceBasename.substring(0, splitAt) + fileNumber;
```
Comment (arthuro555) [inline newIDE/app/src/ResourcesList/ResourceUtils.js:95]: calling `resourceBasename.substring(0, splitAt)` at every iteration seems unnecessary and to add complexity, making the whole thing run longer. You can just make a copy of fileName before the iteration to minimize the method calls:
```suggestion
      const originalFileName = fileName;

      for (let itr = 0; itr < AllResourcesNames.size(); itr++) {
        if (fileName + fileExtension === AllResourcesNames.at(itr)) {
          fileName = roriginalFileName + fileNumber;
```

**Review by arthuro555 — COMMENTED** [review]
Thank you for the PR and contributing again to GDevelop!
I've left a few comments that may help improve the quality of the PR.
Have a good day!

**Comment by rs4231199** [comment]
> Can you explain a bit how this work, how you tested this, in which situation this helps? :)

@4ian I implemented the solution suggested in issue #2171. There is a for-loop that appends numbers to the original filename until it finds a unique name. For example, if tree.png is already present in the project folder and a user imports a file with the name tree.png then it would be added as tree1.png to the folder. If the user imports another file with the name tree.png then it would be added as tree2.png.

I ran checks like suggested in the [development setup readme](https://github.com/4ian/GDevelop/blob/master/newIDE/README.md#tests). I also checked by adding files to the project like the issue's author did in the video.

**Comment by rs4231199** [comment]
@arthuro555 Thanks for the suggestions. I will update changes now.

File: newIDE/app/src/ResourcesList/ResourceUtils.js [hunk 5]
```diff
@@ -80,8 +80,24 @@ export const copyAllToProjectFolder = (
...
+      let fileName = originalFileName,
+        fileNumber = 1;
+
+      for (let itr = 0; itr < allResourcesNames.size(); itr++) {
+        if (fileName + fileExtension === allResourcesNames.at(itr)) {
+          fileName = originalFileName + fileNumber;
+          fileNumber++;
+          itr = 0;
+        }
+      }
```
Comment (4ian) [inline newIDE/app/src/ResourcesList/ResourceUtils.js:98]: Can you investigate using `newNameGenerator`? Search for it in the rest of the codebase.

File: newIDE/app/src/ResourcesList/ResourceUtils.js [hunk 6]
```diff
@@ -80,8 +80,24 @@ export const copyAllToProjectFolder = (
         return resourcePath;
       }
 
-      const resourceBasename = path.basename(resourcePath);
-      const resourceNewPath = path.join(projectPath, resourceBasename);
+      const resourceBasename = path.basename(resourcePath),
+        splitAt = resourceBasename.lastIndexOf('.'),
+        originalFileName = resourceBasename.substring(0, splitAt),
+        fileExtension = resourceBasename.substring(splitAt),
```
Comment (4ian) [inline newIDE/app/src/ResourcesList/ResourceUtils.js:86]: Use `path.extname` to get an extension.

**Comment by rs4231199** [comment]
@4ian Search for `newNameGenerator` in the codebase was helpful. I was able to remove for-loop and improve readability of the code.

**Comment by 4ian** [comment]
That looks good! Can you give it a try to these: 
- importing from another folder a bunch of files that have no conflict in filenames, check that names are preserved.
- importing from another folder a bunch of files that have conflicts in filenames, check that names are updated.

To ensure it's working well :) 
Thanks!

**Comment by rs4231199** [comment]
It is working as expected. 
First I imported 4 files(`Blood1.png`, `Blood2.png`, `Blood3.png`, `BloodHand.png`) and then I imported 5 files(`Blood1.png`, `Blood2.png`, `Blood3.png`, `BloodHand.png`, `BTNClose.png`).
This is final folder content.
```
Blood1.png   Blood2.png   Blood3.png   BloodHand.png 
Blood12.png  Blood22.png  Blood32.png  BloodHand2.png  BTNClose.png
```
After this I imported `Blood.png` twice. It got added with the names `Blood.png` and `Blood4.png`.

**Comment by 4ian** [comment]
Thanks for the test, that sounds great 👍

Merging now, thanks for working on this!

**Pull request merged**